### PR TITLE
fix(array done functions pt. 1)

### PR DESCRIPTION
**Pull request by Spencer6497**
This PR adds done functions to test whether every array item contains completed entries. Part 1.

- **add childDone array helper and add tests**
- **add ausgabenDone for BerH**
- **write array doneFunctions for PKH ausgaben**

### CODE REVIEW

File: app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts [hunk 1]
```diff
@@ -147,16 +151,61 @@ export const ausgabenZusammenfassungDone: ProzesskostenhilfeFinanzielleAngabenGu
...
   context,
-}) => arrayIsNonEmpty(context.ratenzahlungen);
+}) =>
+  arrayIsNonEmpty(context.ratenzahlungen) &&
+  context.ratenzahlungen.every(ratenzahlungDone);
+
+export const sonstigeAusgabeDone = (
+  sonstigeAusgabe: Partial<
+    z.infer<typeof prozesskostenhilfeFinanzielleAngabenContext.sonstigeAusgaben>
+  >[0],
```
Comment (chohner) [inline app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts:195]: this is just `ProzesskostenhilfeFinanzielleAngabenContext['sonstigeAusgaben'][0]`, right? no need to infer again

File: app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts [hunk 2]
```diff
@@ -147,16 +151,61 @@ export const ausgabenZusammenfassungDone: ProzesskostenhilfeFinanzielleAngabenGu
...
+) =>
+  !!sonstigeAusgabe &&
+  sonstigeAusgabe.art !== undefined &&
+  sonstigeAusgabe.zahlungsempfaenger !== undefined &&
+  sonstigeAusgabe.zahlungspflichtiger !== undefined &&
+  (sonstigeAusgabe.zahlungspflichtiger === "myself" ||
+    sonstigeAusgabe.betragEigenerAnteil !== undefined) &&
+  sonstigeAusgabe.betragGesamt !== undefined;
```
Comment (chohner) [inline app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts:203]: in this case, couldn't we already use zod? `return prozesskostenhilfeFinanzielleAngabenContext.sonstigeAusgaben.safeParse(sonstigeAusgabe).success` should do what we want here

File: app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts [hunk 3]
```diff
@@ -147,16 +151,61 @@ export const ausgabenZusammenfassungDone: ProzesskostenhilfeFinanzielleAngabenGu
...
 export const hasVersicherungDone: ProzesskostenhilfeFinanzielleAngabenGuard = ({
   context,
-}) => arrayIsNonEmpty(context.versicherungen);
+}) =>
+  arrayIsNonEmpty(context.versicherungen) &&
+  context.versicherungen.every(versicherungDone);
+
+export const ratenzahlungDone = (
+  ratenzahlung: Partial<
+    z.infer<typeof prozesskostenhilfeFinanzielleAngabenContext.ratenzahlungen>
```
Comment (chohner) [inline app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts:173]: `ProzesskostenhilfeFinanzielleAngabenContext['ratenzahlungen'][0]`

File: app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts [hunk 4]
```diff
@@ -147,16 +151,61 @@ export const ausgabenZusammenfassungDone: ProzesskostenhilfeFinanzielleAngabenGu
     hasRatenzahlungDone({ context }) ||
     hasSonstigeAusgabeDone({ context });
 
+export const versicherungDone = (
+  versicherung: z.infer<
+    typeof prozesskostenhilfeFinanzielleAngabenContext.versicherungen
```
Comment (chohner) [inline app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts:156]: `ProzesskostenhilfeFinanzielleAngabenContext['versicherungen'][0]`

File: app/flows/prozesskostenhilfeFormular/finanzielleAngaben/__test__/doneFunctions.test.ts [hunk 5]
```diff
@@ -76,6 +105,267 @@ describe("Finanzielle Angaben doneFunctions", () => {
...
+        hasRatenzahlungDone({
+          context: {
+            ratenzahlungen: [mockedCompleteRatenzahlung],
+          },
+        }),
+      ).toBe(true);
+    });
+  });
+
+  describe("sonstigeAusgabeDone", () => {
```
Comment (chohner) [inline app/flows/prozesskostenhilfeFormular/finanzielleAngaben/__test__/doneFunctions.test.ts:306]: if we use zod schema we could vastly reduce the tests imo

File: app/flows/beratungshilfeFormular/finanzielleAngaben/doneFunctions.ts [hunk 6]
```diff
@@ -72,13 +74,27 @@ export const andereUnterhaltszahlungenDone: BeratungshilfeFinanzielleAngabenGuar
     context.hasWeitereUnterhaltszahlungen == "no" ||
     arrayIsNonEmpty(context.unterhaltszahlungen);
 
+export const ausgabeDone = (
+  ausgabe: Partial<
+    z.infer<typeof beratungshilfeFinanzielleAngaben.ausgaben>[0]
```
Comment (chohner) [inline app/flows/beratungshilfeFormular/finanzielleAngaben/doneFunctions.ts:79]: `BeratungshilfeFinanzielleAngaben['ausgaben'][0]`

**Review by chohner — COMMENTED** [review]
I think we could remove all the `z.infer()`

File: app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts (hunk 2 above)
Comment (Spencer6497) [reply app/flows/prozesskostenhilfeFormular/finanzielleAngaben/doneFunctions.ts:203]: unfortunately not, as there's no way to tell zod that `betragEigenerAnteil` should be defined _only_ if `zahlungspflichtiger` isn't `myself`

**Pull request merged**